### PR TITLE
feat: update events for credential SNS

### PR DESCRIPTION
**Pull request by ReubenRobertsHEE**
TODO: filter by field updates that actually affect credentials.

TIS21-4150

### CODE REVIEW

**Review by Judge40 — APPROVED** [review]
Looks good, I'm not sure we need to bother filtering the data on this end. The sync service shouldn't/doesn't need to know about credential contents.

File: src/main/java/uk/nhs/hee/tis/trainee/sync/service/TcsSyncService.java [hunk 1]
```diff
@@ -84,6 +84,9 @@ public class TcsSyncService implements SyncService {
   private static final String REQUIRED_ROLE = "DR in Training";
   private static final String[] REQUIRED_NOT_ROLES = {"Placeholder", "Dummy Record"};
 
+  private static final Operation[] UPDATE_OPERATIONS
+      = {Operation.UPDATE, Operation.INSERT, Operation.LOAD};
```
Comment (ReubenRobertsHEE) [inline src/main/java/uk/nhs/hee/tis/trainee/sync/service/TcsSyncService.java:88]: Looking at the operations in the existing SQS messages, it seems we need to monitor all of these.

File: src/main/java/uk/nhs/hee/tis/trainee/sync/service/TcsSyncService.java [hunk 2]
```diff
@@ -162,28 +165,36 @@ public void syncRecord(Record recrd) {
...
-        JsonNode eventJson = objectMapper.valueToTree(Map.of(
+        eventJson = objectMapper.valueToTree(Map.of(
+            "tisId", recrd.getTisId(),
+            "timestamp", timestamp
+        ));
+      } else if (Arrays.asList(UPDATE_OPERATIONS).contains(recrd.getOperation())) {
+        //the credential service can determine if updated fields actually affect issued credentials
+        eventJson = objectMapper.valueToTree(Map.of(
             "tisId", recrd.getTisId(),
+            "record", recrd,
```
[the remark this answers is not in the document]
Comment (Judge40) [reply src/main/java/uk/nhs/hee/tis/trainee/sync/service/TcsSyncService.java:189]: An alternative to the `UPDATE_OPERATIONS.contains(x)` approach would be to use an enhanced switch expression here

```java
JsonNode eventJson = switch(recrd.getOperation()) {
  case DELETE -> objectMapper.valueToTree(Map.of(
      "tisId", recrd.getTisId(),
      "timestamp", timestamp
  ));
  case INSERT, LOAD, UPDATE -> objectMapper.valueToTree(Map.of(
      "tisId", recrd.getTisId(),
      "record", recrd,
      "timestamp", timestamp
  ));
  default -> null;
};
```

Duplication could be reduced a little further by building the JSON separately

```java
Map<String, Object> treeValues = switch(recrd.getOperation()) {
  case DELETE -> Map.of(
      "tisId", recrd.getTisId(),
      "timestamp", timestamp
  );
  case INSERT, LOAD, UPDATE -> Map.of(
      "tisId", recrd.getTisId(),
      "record", recrd,
      "timestamp", timestamp
  );
  default -> null;
};

if (treeValues != null) {
  JsonNode eventJson = objectMapper.valueToTree(treeValues);
  request = new PublishRequest()
      .withMessage(eventJson.toString())
      .withTopicArn(snsTopic);
}
```

File: src/main/java/uk/nhs/hee/tis/trainee/sync/service/TcsSyncService.java [hunk 3]
```diff
@@ -200,16 +211,28 @@ public void publishDetailsChangeEvent(Record recrd) {
...
+   * @return The SNS topic ARN, or null if the table changes are not broadcast.
    */
-  String tableToSnsTopic(String table) {
-    return switch (table) {
-      case TABLE_PLACEMENT -> eventNotificationProperties.deletePlacementEvent();
-      case TABLE_PROGRAMME_MEMBERSHIP, TABLE_CURRICULUM_MEMBERSHIP ->
-          eventNotificationProperties.deleteProgrammeMembershipEvent();
-      default -> null;
-    };
+  String tableToSnsTopic(String table, Operation operation) {
```
Comment (Judge40) [inline src/main/java/uk/nhs/hee/tis/trainee/sync/service/TcsSyncService.java:218]: Should this be private?

File: src/main/java/uk/nhs/hee/tis/trainee/sync/service/TcsSyncService.java [hunk 4]
```diff
@@ -200,16 +211,28 @@ public void publishDetailsChangeEvent(Record recrd) {
...
+    } else if (Arrays.asList(UPDATE_OPERATIONS).contains(operation)) {
+      return switch (table) {
+        case TABLE_PLACEMENT -> eventNotificationProperties.updatePlacementEvent();
+        case TABLE_PROGRAMME_MEMBERSHIP, TABLE_CURRICULUM_MEMBERSHIP ->
+            eventNotificationProperties.updateProgrammeMembershipEvent();
+        default -> null;
+      };
+    } else {
+      return null; //other operations are ignored
+    }
```
Comment (Judge40) [inline src/main/java/uk/nhs/hee/tis/trainee/sync/service/TcsSyncService.java:235]: I'm debating whether it's a more/less readable approach myself, but we could use nested switch statements here.

```java
return switch (operation) {
  case DELETE -> switch (table) {
    case TABLE_PLACEMENT -> eventNotificationProperties.deletePlacementEvent();
    case TABLE_PROGRAMME_MEMBERSHIP, TABLE_CURRICULUM_MEMBERSHIP ->
        eventNotificationProperties.deleteProgrammeMembershipEvent();
    default -> null;
  };
  case INSERT, LOAD, UPDATE -> switch (table) {
    case TABLE_PLACEMENT -> eventNotificationProperties.updatePlacementEvent();
    case TABLE_PROGRAMME_MEMBERSHIP, TABLE_CURRICULUM_MEMBERSHIP ->
        eventNotificationProperties.updateProgrammeMembershipEvent();
    default -> null;
  };
  default -> null;
};
```

File: src/test/java/uk/nhs/hee/tis/trainee/sync/service/TcsSyncServiceTest.java [hunk 5]
```diff
@@ -570,12 +582,57 @@ void shouldIssueEventForSpecifiedTablesWhenOperationDelete(String table)
...
+
+    verifyNoMoreInteractions(snsService);
+  }
+
+  /**
+   * Provide the cartesian product of the tables and update operations that should trigger events.
+   *
+   * @return The stream of arguments.
+   */
+  private static Stream<Arguments> provideUpdateParameters() {
```
[the remark this answers is not in the document]
Comment (Judge40) [reply src/test/java/uk/nhs/hee/tis/trainee/sync/service/TcsSyncServiceTest.java:595]: I'm not convinced it's better, but you could do something like this with streams 😵 
```java
private static Stream<Arguments> provideUpdateParameters() {
  return Stream.of(UPDATE, LOAD, INSERT).flatMap(operation ->
      Stream.of(TABLE_PLACEMENT, TABLE_PROGRAMME_MEMBERSHIP, TABLE_CURRICULUM_MEMBERSHIP)
          .flatMap(table -> Stream.of(
              Arguments.of(operation, table)
          ))
  );
}
```

File: src/main/java/uk/nhs/hee/tis/trainee/sync/service/TcsSyncService.java (hunk 1 above)
Comment (ReubenRobertsHEE) [reply src/main/java/uk/nhs/hee/tis/trainee/sync/service/TcsSyncService.java:88]: Yes, that's right - that's what you get when you let Sonar 'fix' the issue without thinking about actually fixing the issue 🤦

**Comment by ReubenRobertsHEE** [comment]
I give up on the indentation code smells 😂 . This is with the intellij code-formatting. If I manually adjust them to what sonar wants, it seems to change its mind and prefer them the way they were!

**Comment by Judge40** [comment]
> I give up on the indentation code smells 😂 . This is with the intellij code-formatting. If I manually adjust them to what sonar wants, it seems to change its mind and prefer them the way they were!

Have you imported the Google formatter?
https://github.com/google/styleguide/blob/gh-pages/intellij-java-google-style.xml

**Comment by ReubenRobertsHEE** [comment]
> > I give up on the indentation code smells 😂 . This is with the intellij code-formatting. If I manually adjust them to what sonar wants, it seems to change its mind and prefer them the way they were!
> 
> Have you imported the Google formatter? https://github.com/google/styleguide/blob/gh-pages/intellij-java-google-style.xml

Ah cool, I have now. (It fixes the alignment of line-broken parameters in a function, but makes no difference to this specific issue) 👍

**Pull request merged**